modules/distributor/usage/tracker.go

```diff
@@ -0,0 +1,425 @@
+		// it performs the floating point math once on the total, instead of accumulating 1/N + 1/N ... errors.
+		batchPortion := int(math.RoundToEven(float64(unaccountedForBatchData) / float64(totalSpanCount)))
+
+		// To account for the accumulated error we dump the remaining delta onto the first span, which can be negative.
+		// The result ensures the total recorded bytes matches the input.
+		firstSpanPortion := unaccountedForBatchData - batchPortion*totalSpanCount
+
+		// Reset value buffer for every batch.
+		for k := range buffer1 {
+			buffer1[k] = missingLabel
+		}
+
+		if batch.Resource != nil {
+			for _, a := range batch.Resource.Attributes {
+				v := a.Value.GetStringValue()
+				if v == "" {
+					continue
+				}
+				for _, m := range mapping {
+					if m.from == a.Key {
+						buffer1[m.to] = v
+						break
+					}
+				}
+			}
+		}
+
+		var bucket *bucket
+
+		for i, ss := range batch.ScopeSpans {
+			for j, s := range ss.Spans {
+				sz := s.Size()
+				sz += protoLengthMath(sz)
+				sz += batchPortion // Incrementally add 1/Nth worth of the unaccounted for batch data
+				if i == 0 && j == 0 {
+					sz += firstSpanPortion
+				}
+
+				// Reset to batch values to for some spans having missing values.
+				copy(buffer2, buffer1)
+
+				for _, a := range s.Attributes {
+					v := a.Value.GetStringValue()
+					if v == "" {
+						continue
+					}
+					for _, m := range mapping {
+						if m.from == a.Key {
+							buffer2[m.to] = v
+							break
+						}
+					}
+				}
+
+				// Every span can be a different series.
+				// If the values buffer hasn't changed then we
+				// know it's the same bucket and avoid hashing again.
+				// This shows up in 2 common cases:
+				//  - Dimensions are only resource attributes
+				//  - Runs of spans with the same attributes
+				// NOTE - Not happy about the slices.Equal to detect when
+				// to rehash, but couldn't figure out a better way for now.
+				// The difficulty is tracking bucket dirty status while
+				// resetting to batch values and recording the span values.
+				if bucket == nil || !slices.Equal(buffer2, last) {
+					bucket = data.getSeries(buffer2, max)
+					copy(last, buffer2)
+				}
+				bucket.Inc(uint64(sz), now)
+			}
+		}
+	}
+}
+
+func (u *Tracker) PurgeRoutine() {
+	purge := time.NewTicker(defaultPurgePeriod)
+	for range purge.C {
+		u.purge()
+	}
+}
+
+func (u *Tracker) purge() {
+	u.mtx.Lock()
+	defer u.mtx.Unlock()
+
+	stale := time.Now().Add(-u.cfg.StaleDuration).Unix()
+
+	for t, data := range u.tenants {
+		for h, s := range data.series {
+			if s.lastUpdated <= stale {
+				delete(data.series, h)
+			}
+		}
+
+		if len(data.series) == 0 {
+			// Remove empty tenant
+			delete(u.tenants, t)
+		}
+	}
+}
+
+func (u *Tracker) Handler() http.Handler {
+	return promhttp.HandlerFor(u.reg, promhttp.HandlerOpts{})
+}
+
+func (u *Tracker) Describe(chan<- *prometheus.Desc) {
```

Comment:
Do we want to write the `desc` here?

Reply:
Good question. The short answer is no.  `Describe()` is called during [registration](https://github.com/prometheus/client_golang/blob/main/prometheus/registry.go#L279) on startup, so the list of tenants and series will always be empty.  Within there it mentions that a collector with no descriptions is an "unchecked" collector.  Unsure if there are any notable considerations from that.  It does work as-is.